server/src/main/java/org/opensearch/index/store/Store.java

```diff
@@ -187,15 +188,36 @@ protected void closeInternal() {
     };
 
     public Store(ShardId shardId, IndexSettings indexSettings, Directory directory, ShardLock shardLock) {
-        this(shardId, indexSettings, directory, shardLock, OnClose.EMPTY);
+        this(shardId, indexSettings, directory, null, shardLock, OnClose.EMPTY);
+    }
+
+    public Store(ShardId shardId, IndexSettings indexSettings, Directory directory, Directory secondaryDirectory, ShardLock shardLock) {
+        this(shardId, indexSettings, directory, secondaryDirectory, shardLock, OnClose.EMPTY);
     }
 
     public Store(ShardId shardId, IndexSettings indexSettings, Directory directory, ShardLock shardLock, OnClose onClose) {
+        this(shardId, indexSettings, directory, null, shardLock, onClose);
+    }
+
+    public Store(
+        ShardId shardId,
+        IndexSettings indexSettings,
+        Directory directory,
+        Directory secondaryDirectory,
+        ShardLock shardLock,
+        OnClose onClose
+    ) {
         super(shardId, indexSettings);
         final TimeValue refreshInterval = indexSettings.getValue(INDEX_STORE_STATS_REFRESH_INTERVAL_SETTING);
         logger.debug("store stats are refreshed with refresh_interval [{}]", refreshInterval);
         ByteSizeCachingDirectory sizeCachingDir = new ByteSizeCachingDirectory(directory, refreshInterval);
         this.directory = new StoreDirectory(sizeCachingDir, Loggers.getLogger("index.store.deletes", shardId));
+        if (secondaryDirectory != null) {
+            ByteSizeCachingDirectory sizeCachingSecondaryDir = new ByteSizeCachingDirectory(secondaryDirectory, refreshInterval);
+            this.secondaryDirectory = new StoreDirectory(sizeCachingSecondaryDir, Loggers.getLogger("index.store.deletes", shardId));
+        } else {
+            this.secondaryDirectory = null;
+        }
```

Comment:
I think we should we have a Directory factory instead instead of always having two directories always. Primary and secondary directory sounds confusing as well

Reply:
The presence of branching logic is primary / secondary is confusing, please consider subclassing the `Store` (fe `CachedStore`) to have a clean separation between regular `Store` and `Store` backed by a cache / secondary.

Reply:
As I think more on the comments from @Bukhtawar and @eirsep, we may need to have a `CompositeDirectory` which will have 2 directories (we can create compositions for multiple directories if required). `CompositeDirectory` will be abstract and its implementations can define how to handle operations on the encapsulated directories. Implementation can be parallel or sequential.

Reply:
As I think more, going either with `CompositeStore` or `CompositeDirectory` would be too early. I was able to get a quick implementation out with [CompositeDirectory](https://github.com/sachinpkale/OpenSearch/commit/0647e2b45211cf28f63f1ae538de587b1a46a871) but it would require most of the RemoteDirectory methods to be implemented before we can actually start using it. Using these abstractions will not avoid having a reconciliation mechanism between local and remote directory. I made some changes in the latest commit and `Store` is not changed in the latest change. Please take a look.